### ISSUE TITLE
w3wp crashes with bucketed content

### ISSUE THREAD

**Issue by blipson89** [kamsar/Unicorn#177]
When trying to add an item to an item bucket, the w3wp process crashes with the below stack trace. Per our discussion, you suspect it has something to do with Unicorn/Rainbow compiled against SC 8.2, but the site I'm working with is SC 8.1
```
13676 13:28:18 ERROR Unhandled exception detected. The ASP.NET worker process will be terminated. 
Exception: System.TypeLoadException
Message: Could not load type 'Sitecore.Caching.Generics.CustomCache`1' from assembly 'Sitecore.Kernel, Version=8.1.0.0, Culture=neutral, PublicKeyToken=null'.
Source: Unicorn
   at Unicorn.Data.DataProvider.UnicornDataProvider.RemoveItemFromCaches(IItemMetadata metadata, String databaseName)
   at Rainbow.Storage.SerializationFileSystemDataStore.<>c__DisplayClass23_0.<CreateTree>b__0(IItemMetadata metadata)
   at Rainbow.Storage.SerializationFileSystemTree.HandleDataItemChanged(String path, TreeWatcherChangeType changeType)
   at Rainbow.Storage.TreeWatcher.FlushEvents(Object ignored)
   at System.Threading.ExecutionContext.RunInternal(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.ExecutionContext.Run(ExecutionContext executionContext, ContextCallback callback, Object state, Boolean preserveSyncCtx)
   at System.Threading.TimerQueueTimer.CallCallback()
   at System.Threading.TimerQueueTimer.Fire()
   at System.Threading.TimerQueue.FireNextTimers()
```

Unicorn 3.3.0-RC4
Rainbow 1.4.0-RC07
Sitecore 8.1

**Comment by blipson89** [comment]
Update: I tried rebuilding Unicorn and Rainbow to use the 8.1 DLL and I'm still running into this issue.

**Comment by kamsar** [comment]
This is caused by a type definition mismatch - in 8.2 the `Database.DataCache` and `Database.ItemPathsCache` properties get their `Clear()` method from `CustomCache<T>` and in 8.1 their `Clear()` methods are defined on their own classes.

This causes an error when Unicorn, built against 8.2 DLLs, tries to make the cache clear call on an object that does not exist in the 8.1 assembly.

**Comment by kamsar** [comment]
This issue solved by https://github.com/kamsar/Unicorn/commit/185fdc0313ca971f397f22b8a03cd663867283da; relatedly extraneous cache clearing on this type of operation fixed by https://github.com/kamsar/Rainbow/commit/03dcd96fb4a1390e100bc296c581bd28730a68b6

**Comment by kamsar** [comment]
Released in 3.3.1
